src/cdk/overlay/mouse/overlay-mouse-click-dispatcher.ts

```diff
@@ -0,0 +1,96 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {DOCUMENT} from '@angular/common';
+import {Inject, Injectable, OnDestroy} from '@angular/core';
+import {OverlayRef} from '../overlay-ref';
+
+/**
+ * Service for dispatching mouse click events that land on the body to appropriate overlay ref,
+ * if any. It maintains a list of attached overlays to determine best suited overlay based
+ * on event target and order of overlay opens.
+ */
+@Injectable({providedIn: 'root'})
+export class OverlayMouseClickDispatcher implements OnDestroy {
```

Comment:
It looks like this class shares most of its code with the keyboard dispatcher. I think that we should only have one dispatcher that deals with both so we don't have to duplicate all of the code. If we rolled this into the keyboard dispatcher the class name won't make sense anymore, but we can rename it in a backwards-compatible way.

Reply:
Agreed that, if they share a lot of code, we should combine the two classes. This name also says "Mouse"- does this apply for touch event as well?

Reply:
@crisbeto and @jelbourn you are right. I was looking at `OverlayKeyboardDispatcher` when I was implementing `OverlayClickDispatcher`. However looking at the code if we combine both these in one single class it will be a little awkward. On `add` `OverlayKeyboardDispatcher` adds event listener for `keydown` while `OverlayClickDispatcher` adds one for `click`. Then in the event listener function both dispatchers implements different logic. If I put both in one dispatcher it should have a lot of `ifs` inside. Additionally if in future a need for another dispatcher arise we should add more `ifs`. This is why I’ve extracted the common logic in base `OverlayBaseDispatcher` class which both keyboard and click dispatchers extend. What you think about this?